File: libs/go/types/business/delegation.go
package business

// DelegationStruct represents the delegation data structure
// This structure matches the format used by MetaMask delegation toolkit
// Reference: https://docs.metamask.io/delegation-toolkit/how-to/create-delegation/
type DelegationStruct struct {
	Delegate  string         `json:"delegate"`  // The address being delegated to
	Delegator string         `json:"delegator"` // The address creating the delegation
	Authority string         `json:"authority"` // Hex-encoded authority (typically the delegator)
	Caveats   []CaveatStruct `json:"caveats"`   // Restrictions on the delegation
	Salt      string         `json:"salt"`      // Random value for uniqueness
	Signature string         `json:"signature"` // The delegation signature
}

// AuthorityStruct represents the authority information in a delegation
// Based on the working integration test format
type AuthorityStruct struct {
	Scheme    string `json:"scheme"`
	Signature string `json:"signature"`
	Signer    string `json:"signer"`
}

// CaveatStruct represents a single caveat in a delegation
// Based on MetaMask delegation toolkit: https://docs.metamask.io/delegation-toolkit/concepts/caveat-enforcers/
type CaveatStruct struct {
	Enforcer string `json:"enforcer"` // Address of the caveat enforcer contract
	Terms    string `json:"terms"`    // Encoded parameters defining the specific restrictions (hex string)
}

File: libs/ts/delegation/src/types/delegation.ts
import { type Hex, type Address } from 'viem';

/**
 * Authority structure for delegation
 * Based on the working integration test format
 */
export interface AuthorityStruct {
  scheme: string;
  signature: string;
  signer: string;
}

/**
 * Caveat structure for delegation restrictions
 * Based on MetaMask delegation toolkit: https://docs.metamask.io/delegation-toolkit/concepts/caveat-enforcers/
 */
export interface CaveatStruct {
  enforcer: Address; // Address of the caveat enforcer contract
  terms: Hex;        // Encoded parameters defining the specific restrictions
}

/**
 * Complete delegation structure
 * This matches the format used by both Go backend and TypeScript frontend
 * Compatible with MetaMask delegation toolkit
 */
export interface DelegationStruct {
  delegate: Address;
  delegator: Address;
  authority: AuthorityStruct;
  caveats: CaveatStruct[];
  salt: string;
  signature: Hex;
}

/**
 * MetaMask delegation toolkit types re-export
 * For direct compatibility with existing code
 */
export type { Delegation, Caveat } from '@metamask/delegation-toolkit';

/**
 * Network configuration interface
 */
export interface NetworkConfig {
  chainId: number;
  name: string;
  rpcUrl: string;
  bundlerUrl?: string;
  pimlicoApiKey?: string;
  blockExplorer?: string;
  nativeCurrency: {
    name: string;
    symbol: string;
    decimals: number;
  };
}

/**
 * Validation result for delegation validation
 */
export interface ValidationResult {
  isValid: boolean;
  errors: string[];
}